src/core/ext/filters/client_channel/resolver/binder/binder_resolver.cc

```diff
@@ -0,0 +1,149 @@
+#include "src/core/ext/filters/client_channel/server_address.h"
+#include "src/core/lib/address_utils/parse_address.h"
+#include "src/core/lib/channel/channel_args.h"
+#include "src/core/lib/gpr/string.h"
+
+namespace grpc_core {
+namespace {
+
+class BinderResolver : public Resolver {
+ public:
+  BinderResolver(ServerAddressList addresses, ResolverArgs args)
+      : result_handler_(std::move(args.result_handler)),
+        addresses_(std::move(addresses)),
+        channel_args_(grpc_channel_args_copy(args.args)) {}
+
+  ~BinderResolver() override { grpc_channel_args_destroy(channel_args_); };
+
+  void StartLocked() override {
+    Result result;
+    result.addresses = std::move(addresses_);
+    result.args = channel_args_;
+    channel_args_ = nullptr;
+    result_handler_->ReturnResult(std::move(result));
+  }
+
+  void ShutdownLocked() override {}
+
+ private:
+  std::unique_ptr<ResultHandler> result_handler_;
+  ServerAddressList addresses_;
+  const grpc_channel_args* channel_args_ = nullptr;
+};
+
+class BinderResolverFactory : public ResolverFactory {
+ public:
+  bool IsValidUri(const URI& uri) const override {
+    return ParseUri(uri, nullptr);
+  }
+
+  OrphanablePtr<Resolver> CreateResolver(ResolverArgs args) const override {
+    ServerAddressList addresses;
+    if (!ParseUri(args.uri, &addresses)) return nullptr;
+    return MakeOrphanable<BinderResolver>(std::move(addresses),
+                                          std::move(args));
+  }
+
+  const char* scheme() const override { return "binder"; }
+
+ private:
+  static grpc_error_handle BinderAddrPopulate(
+      std::string path, grpc_resolved_address* resolved_addr) {
+    // strip off leading '/' characters
+    size_t p = path.find_first_not_of('/');
+    if (p == std::string::npos) {
+      path = "";
+    } else {
+      path = path.substr(p);
+    }
+    if (path.empty()) {
+      return GRPC_ERROR_CREATE_FROM_CPP_STRING("path is empty");
+    }
+    // Store parsed path in a unix socket so it can be reinterpreted as
+    // sockaddr. An invalid address family (AF_MAX) is set to make sure it won't
+    // be accidentally used.
+    memset(resolved_addr, 0, sizeof(*resolved_addr));
+    struct sockaddr_un* un =
+        reinterpret_cast<struct sockaddr_un*>(resolved_addr->addr);
+    un->sun_family = AF_MAX;
+    static_assert(sizeof(un->sun_path) >= 101,
+                  "unix socket path size is unexpectedly short");
+    if (path.size() + 1 > sizeof(un->sun_path)) {
+      return GRPC_ERROR_CREATE_FROM_CPP_STRING(path +
+                                               " is too long to be handled");
+    }
+    strcpy(un->sun_path, path.c_str());
+    resolved_addr->len =
+        static_cast<socklen_t>(sizeof(un->sun_family) + path.size() + 1);
+    return GRPC_ERROR_NONE;
+  }
+
+  static bool ParseUri(const URI& uri, ServerAddressList* addresses) {
+    grpc_resolved_address addr;
+    {
+      if (uri.scheme() != "binder") {
```

Comment:
This check is unnecessary; this factory won't be selected in the first place if that scheme isn't used.